### ISSUE TITLE
False positive scala compile errors

### ISSUE THREAD

**Issue by chris-martin** [AtomLinter/linter-scalac#52]
[This demo project](https://github.com/chris-martin/atom-lint-scala-demo) compiles successfully with SBT, but Atom shows errors.

![screenshot from 2016-07-04 18-45-01](https://cloud.githubusercontent.com/assets/399718/16570630/70c1e0d8-4217-11e6-8b19-f5be5e6280c7.png)

**Comment by andystanton** [comment]
Hi Chris, thanks for the issue report!

This linter is dumb in that it just uses scalac to lint the source code. Scalac doesn't know about sbt dependencies so won't compile against them unless explicitly told.

I think the intention of this project was to be able to lint a single source file or two rather than entire sbt projects, although there is a way to tell linter-scalac what extra libraries you want to compile against described here: https://github.com/AtomLinter/linter-scalac#classpath.

There is an Atom package called [sbt](https://atom.io/packages/sbt) that does lint based on sbt output but requires you to invoke sbt compile from the accompanying terminal rather than on save/on the fly.

You could adapt that project, or this linter, or make your own one that allowed for automatic linting.

Other alternatives you might consider for Scala development:

* IntelliJ IDEA Community Edition (my personal recommendation)
* [Scala IDE](http://scala-ide.org)

I'm closing this issue as there is no plan to add sbt linting to this project.

**Comment by chris-martin** [comment]
Hi @andystanton - Thanks for the response. I think there is more to this than meets the eye, though.

I did generate a `.classpath` file, which is why there are no errors marked on the import or use of the `ActorSystem` constructor. The dependencies are being recognized, and the majority of my project is linting correctly. There just seems to be a peculiarity with those couple lines.

**Comment by chris-martin** [comment]
I wonder if it's possible something's using the wrong version of akka-core, because `terminate` and `whenTerminated` were only added to `ActorSystem` in the latest version `2.4.x`.

**Comment by andystanton** [comment]
Apologies for closing this prematurely. Your suspicions are well founded.

I ran this command for your project, which is essentially what is invoked by the linter (except the classpath is taken from .classpath):

```sh
scalac -verbose -classpath "$(sbt 'export fullClasspath' | tail -1)" src/main/scala/demo/package.scala
```

The following line occurs in the output:

```
[loaded class file /usr/local/Cellar/scala/2.11.8/libexec/lib/akka-actor_2.11-2.3.10.jar(akka/actor/ActorSystem.class) in 4ms]
```

2.3.10 is clearly not the 2.4.7 specified in sbt, and interestingly appears to be bundled with Scala.

It seems this is known behaviour:

* http://stackoverflow.com/q/19594157
* http://stackoverflow.com/a/36142479

Obviously deleting jars from the Scala distribution isn't a solution, so I'm looking into whether it's possible to get scalac to ignore that particular jar.

**Comment by andystanton** [comment]
I whipped up a hack proof of concept that works. I put this script in the root folder of your example with a populated .classpath file and it compiled fine.

It would be possible to translate this to javascript for the linter, but I'm hoping I can find a less hacky way of doing it.

```
#!/usr/bin/env bash

cp="$(scalac -verbose -help 2>&1 \
  | grep -E '^\[search path for class files: ' \
  | sed -Ee 's/^\[search path for class files: (.*):.]$/\1/')"

function getlibname() {
  sed -Ee 's/.+\/([[:alnum:]-]+)(_[[:digit:]]+\.[[:digit:]]+(\.[[:digit:]]+)?)?(-[[:digit:]]+\.[[:digit:]]+(\.[[:digit:]]+)?)?\.jar/\1/'<<<"${1}"
}

java_cp=""
sbt_cp="$(cat .classpath)"

for element in ${cp//:/ } ; do
  if [[ "${element}" =~ scala ]]; then
    lib_name="$(getlibname "${element}")"
    found=false

    for sbt_element in ${sbt_cp//:/ } ; do
      if [[ "${sbt_element}" =~ \.jar ]]; then
        sbt_lib_name="$(getlibname "${sbt_element}")"
        if [ "${lib_name}" == "${sbt_lib_name}" ]; then
          found=true
        fi
      fi
    done

    if [ "${found}" == "false" ]; then
      java_cp="${java_cp}:${element}"
    fi
  else
    java_cp="${java_cp}:${element}"
  fi
done

java_cp="$(echo "${java_cp}" | cut -c 2-)"

mkdir -p target
scalac -d target -nobootcp -javabootclasspath "${java_cp}" -classpath "${sbt_cp}" "$(find . -name '*.scala')"
```
